Effect of seq depth on gene detection by functional category
========================================================

Load synapse and samples metadata file.
```{r cache=TRUE, warning=FALSE,results='hide'}
library('synapseClient')
synapseLogin()
dataFile = synGet('syn2340130')
metaFile = synGet('syn2299154')
```

```{r cache=TRUE}
geneCounts  = read.delim(dataFile@filePath, header=T, row.names = 1)
metadata = read.csv(getFileLocation(metaFile))
```

```{r warning=FALSE}
setwd('~/Computing/commonmind/analysis/')
source('~/Computing/sage-data-analysis/rnaseq_analysis_functions.R')
```

Make final dataset, with bad samples excluded. 
```{r}
#colnames(metadata)
metadata_freeze = metadata[-which(metadata$DLPFC_RNA_report..Exclude. == 1),]
total_samples = nrow(metadata_freeze)
table(metadata_freeze$DLPFC_RNA_report..Exclude.)
metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID = factor(metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID)

metadata_freeze = metadata_freeze[order(metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID),]
geneCounts_freeze = geneCounts[,which(as.factor(colnames(geneCounts)) %in% metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID)]
geneCounts_freeze = geneCounts_freeze[,order(as.factor(colnames(geneCounts_freeze)))]

head(colnames(geneCounts_freeze))
head(metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID)
tail(colnames(geneCounts_freeze))
tail(metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID)
```


1-Make list of functional categories. I obtained the list from biomart, by asking for all ENSG that have a GO ID.
```{r}
ENSG_GO = read.delim("../../ENSG_GO_mart_export.txt", header = TRUE)
all_GO = unique(ENSG_GO$GO.Term.Accession)

```

2-Run regression on each category.
```{r cache=TRUE}
goBetas = matrix(NA, nrow=length(all_GO),ncol=3)
rownames(goBetas) = all_GO
colnames(goBetas) = c("slope", "intercept", "num_genes")

for (i in 1:length(all_GO)) {
  goGenes = unique(ENSG_GO$Ensembl.Gene.ID[which(ENSG_GO$GO.Term.Accession == all_GO[i])])
  goRownames = which(rownames(geneCounts_freeze) %in% goGenes)
  detected_GO = apply(geneCounts_freeze[goRownames,],MARGIN=2,countDetected)
#  plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), detected_GO/length(goGenes), ylim = range(0,1))
  lm_results = summary(lm(detected_GO/length(goGenes) ~ log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)))$coefficients
  goBetas[i,1] = lm_results[2,1]
  goBetas[i,2] = lm_results[1,1]
  goBetas[i,3] = length(goGenes)
}
```

3-Compare beta estimates for slope, which is the detection/depth relationship.
```{r}
xx = as.list(GOTERM)

hist(goBetas[,1], col = "papayawhip", main = "Detection vs depth\nGO categories", xlab = "regression slope estimate")
plot(goBetas[,2],goBetas[,1], main = "Slope vs Intercept", xlab = "Intercept", ylab = "Slope")
plot(log10(goBetas[,3]),goBetas[,1], main = "Slope vs number of genes", xlab = "number of genes", ylab = "Slope")
```

4-Choose some categories that have highest rate of detection increase with depth, but take into account the number of genes affiliated with each GO group. Essentially, I just drew a line on the graph and looked at things that are above the line.
```{r}
goBetas_reduced = goBetas[-which(goBetas[,3] < 6),]
plot(log10(goBetas_reduced[,3]),goBetas_reduced[,1], main = "Slope vs number of genes", xlab = "log10(number of genes)", ylab = "Slope")
abline(a=0.47,b=-0.16, col = "green", lwd = 2)

range(goBetas_reduced[,3])
goBetas_expected = 0.47-(0.16*log10(as.numeric(goBetas_reduced[,3])))
goBetas_extreme = which(goBetas_reduced[,1] > goBetas_expected)
length(goBetas_extreme)
indicies = which(names(xx) %in% names(goBetas_extreme))
for (i in 1:length(indicies)) {
  print(Term(xx[[indicies[i]]]))
}
```