Detection analysis on gene counts for common mind phase I RNAseq data
========================================================

Load synapse and Ensembl gene counts file.

```{r cache=TRUE, warning=FALSE,results='hide'}
library('synapseClient')
synapseLogin()
dataFile = synGet('syn2340130')
metaFile = synGet('syn2299154')
```

```{r cache=TRUE}
geneCounts  = read.delim(dataFile@filePath, header=T, row.names = 1)
metadata = read.csv(getFileLocation(metaFile))
```

```{r warning=FALSE}
setwd('~/Computing/commonmind/analysis')
source('~/Computing/rnaseq_analysis_functions.R')
```

Which samples are repeated in the input file?

```{r}
which(table(metadata$DLPFC_RNA_isolation..Sample.RNA.ID) > 1)
```

Make final dataset, with bad samples excluded. 

```{r}
#colnames(metadata)
metadata_freeze = metadata[-which(metadata$DLPFC_RNA_report..Exclude. == 1),]
total_samples = nrow(metadata_freeze)
table(metadata_freeze$DLPFC_RNA_report..Exclude.)
metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID = factor(metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID)

metadata_freeze = metadata_freeze[order(metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID),]
geneCounts_freeze = geneCounts[,which(as.factor(colnames(geneCounts)) %in% metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID)]
geneCounts_freeze = geneCounts_freeze[,order(as.factor(colnames(geneCounts_freeze)))]

head(colnames(geneCounts_freeze))
head(metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID)
tail(colnames(geneCounts_freeze))
tail(metadata_freeze$DLPFC_RNA_isolation..Sample.RNA.ID)
```

Detection analysis of biotypes
-----------------------------------------------------------

```{r}
head(rownames(geneCounts_freeze))
geneCounts_freeze_temp = geneCounts_freeze
geneCounts_freeze = addBiotype(geneCounts_freeze_temp)
rm(geneCounts_freeze_temp)
totalBiotypeCounts = table(getBM(attributes=c("ensembl_gene_id", "gene_biotype"), mart=Hs)[,2])
```


```{r echo=FALSE}
plotDetectionByBiotype=function(inBiotypeCounts, inNormCounts){
  par(mfrow = c(3,3))
  corrStats = matrix(0, nrow = nrow(inNormCounts), ncol = 2)
  for (i in 1:27) {
     results = cor.test(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inBiotypeCounts[i,])
     corrStats[i,1] = as.vector(results[[3]])
     corrStats[i,2] = as.vector(results[[4]])
     
    plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inNormCounts[i,], main = rownames(inBiotypeCounts)[i], xlab = "mapped reads", ylab = "fraction detected", ylim = range(0,1))
    lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inNormCounts[i,]), col = "blue", lwd = 2)    
  }
  
  # interesting categories -- this section for plotting selected results from the above detailed ones
#   par(mfrow = c(2,3))
#   interesting = c(2, 10,16,17,18,21)
#   for (i in 1:length(interesting)) {
#     plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inNormCounts[interesting[i],], main = paste(rownames(inBiotypeCounts)[interesting[i]], "\nn=", detectedTotal[interesting[i]], sep = ""), xlab = "mapped reads", ylab = "fraction detected", ylim = range(0,1))
#     lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inNormCounts[interesting[i],]), col = "blue", lwd = 2)
#   }
  par(mfrow = c(1,1))
#  print(corrStats)
  return(corrStats)
}
```

Plot detection per biotype across varying thresholds for minimum read count. Print correlation results and linear regression results.

```{r echo=FALSE}
corrResults = matrix(0, nrow = length(totalBiotypeCounts), ncol = 6)
rownames(corrResults) = names(totalBiotypeCounts)
colnames(corrResults) = rep(c("estimate", "pval"), 3)
lmResults = matrix(0, nrow = length(totalBiotypeCounts), ncol = 6)
rownames(lmResults) = names(totalBiotypeCounts)
colnames(lmResults) = rep(c("estimate", "pval"), 3)
```

```{r}
j = 1
for (i in 2:4) {
  print(paste('Results for minimum count = ', i, sep = ''))
  biotypeCountsPerSampleFilter = apply(geneCounts_freeze[,1:total_samples], 2, countDetectedByBiotype, biotypes = geneCounts_freeze$biotype, filter=i)
  detectedTotal = totalBiotypeCounts[which(names(totalBiotypeCounts) %in% rownames(biotypeCountsPerSampleFilter))]
  normBiotypeCounts = apply(biotypeCountsPerSampleFilter, MARGIN=2, function(x) x/detectedTotal)
  corrResults[1:length(detectedTotal),j:(j+1)] = plotDetectionByBiotype(inBiotypeCounts=biotypeCountsPerSampleFilter,inNormCounts=normBiotypeCounts)
  lmResults[1:length(detectedTotal),j:(j+1)] = regressDetectionOnMetadata(inDetectCounts=normBiotypeCounts, inMetaData=log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads))
  j = j + 2
}
```

```{r}
print(corrResults)
```

Regression analysis of relatinoship between detection and depth per biotype.
```{r}
# sets of two columns: first is beta estimate for slope, second is p-value
print(lmResults)
```

Detection analysis of expression levels, all biotypes together.
---------------------------------------------------------------

```{r fig.height=4, fig.width=4}
gene_lengths = getGeneLengths()
hist(log10(gene_lengths$length), main='Distribution of gene lengths by gene start stop')
```

Looking at control samples only.

```{r echo=c(2:6), fig.height=6, fig.width=6}
library('edgeR')
table(metadata_freeze$Dx)
data.dge = DGEList(counts=geneCounts_freeze[,1:total_samples],group=factor(metadata_freeze$Dx), remove.zeros=TRUE)
data.dge = calcNormFactors(data.dge)
control.dge = data.dge[,which(data.dge$samples$group == "Control")]
control_gene_means = rowMeans(rpkm(control.dge, gene.length=gene_lengths$length[match(rownames(getCounts(control.dge)), gene_lengths$ensembl_gene_id)], log=TRUE,normalized.lib.sizes=TRUE))
```

```{r fig.height=6, fig.width=6}
hist(rpkm(control.dge, gene.length=gene_lengths$length[match(rownames(getCounts(control.dge)), gene_lengths$ensembl_gene_id)],log=TRUE,normalized.lib.sizes=TRUE), main = "control samples, all genes", col = "thistle", xlab = "RPKM")
hist(control_gene_means, main = "gene averages across samples\ncontrol samples only", col = "thistle", breaks = 50)
```

Separate into high/med/low expression datasets.

```{r}
range(control_gene_means, na.rm = TRUE)
lowExpress = which(control_gene_means < -7)
highExpress = which(control_gene_means > -2)
medExpress_temp = which(control_gene_means >= -7)
medExpress = setdiff(medExpress_temp, highExpress)
sum(length(lowExpress), length(highExpress), length(medExpress))
rm(medExpress_temp)

control_high.dge = control.dge[highExpress,]
control_med.dge = control.dge[medExpress,]
control_low.dge = control.dge[lowExpress,]

control_high_detect = apply(getCounts(control_high.dge), MARGIN=2, FUN=countDetected)
control_med_detect = apply(getCounts(control_med.dge), MARGIN=2, FUN=countDetected)
control_low_detect = apply(getCounts(control_low.dge), MARGIN=2, FUN=countDetected)
```

Plot relationships between fraction detected and mapped reads by expression level.

```{r}
#low
plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_low_detect/length(lowExpress), main = paste("detection of low-expression genes", "\nn=", length(lowExpress), sep = ""), xlab = "log10(mapped reads)", ylab = "fraction detected", ylim = range(0,1))
lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_low_detect/length(lowExpress)), col = "blue", lwd = 2)    
cor.test(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_low_detect)

# medium
plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_med_detect/length(medExpress), main = paste("detection of med-expression genes", "\nn=", length(medExpress), sep = ""), xlab = "log10(mapped reads)", ylab = "fraction detected", ylim = range(0,1))
lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_med_detect/length(medExpress)), col = "blue", lwd = 2)    
cor.test(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_med_detect)


# high
plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_high_detect/length(highExpress), main = paste("detection of high-expression genes", "\nn=", length(highExpress), sep = ""), xlab = "log10(mapped reads)", ylab = "fraction detected", ylim = range(0,1))
lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_high_detect/length(highExpress)), col = "blue", lwd = 2)    
cor.test(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_high_detect)
```


What genes are expressed at hi/med/low levels? Do they look like neurons?
--------------------------------------

```{r, echo=FALSE}
## trying to look at categories in each group
# pcGenes = getByBiotype()
# highExpress_ENSGpc = intersect(rownames(getCounts(control_high.dge)), pcGenes[,1])
# medExpress_ENSGpc = intersect(rownames(getCounts(control_med.dge)), pcGenes[,1])
# lowExpress_ENSGpc = intersect(rownames(getCounts(control_low.dge)), pcGenes[,1])
# write.csv(lowExpress_ENSGpc, file = "lowexpress_ENSGpc", quote = FALSE, row.names = FALSE, col.names = FALSE)
# write.csv(highExpress_ENSGpc, file = "highexpress_ENSGpc", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

Getting neuro-relevant genes from GO, then looking at expression level of those genes in our samples.

```{r,echo=FALSE}
printFractionalCounts=function(categoryENSGvec, categoryName){
  print(paste("Number of genes in category", length(categoryENSGvec), sep = " "))
  print(paste("High expression genes as fraction of all genes:", length(highExpress) / nrow(control.dge), sep = " "))
  print(paste("Fraction of", categoryName, "genes found in high expression group:", length(intersect(rownames(getCounts(control_high.dge)), categoryENSGvec))/length(categoryENSGvec), sep = " "))
  
  print(paste("Medium expression genes as fraction of all genes:", length(medExpress) / nrow(control.dge), sep = " "))
  print(paste("Fraction of", categoryName, "genes found in med expression group:", length(intersect(rownames(getCounts(control_med.dge)), categoryENSGvec))/length(categoryENSGvec), sep = " "))
  
  print(paste("Low expression genes as fraction of all genes:", length(lowExpress) / nrow(control.dge), sep = " "))
  print(paste("Fraction of", categoryName, "genes found in low expression group:", length(intersect(rownames(getCounts(control_low.dge)), categoryENSGvec))/length(categoryENSGvec), sep = " "))  
}
```

Genes in neuron-related cellular component GO
```{r cache=TRUE, warning=FALSE}
neuron_part = getGenesForGOOffspring("GO:0097458")
synapse = getGenesForGOOffspring("GO:0045202")
vesicle = getGenesForGOOffspring("GO:1990008") # nothing found
neuron_and_synape = union(neuron_part, synapse)
printFractionalCounts(neuron_and_synape, "neuron-synapse")
```

Genes in neuron birth-growth-death biological process GO.

```{r cache=TRUE}
neurogenesis = getGenesForGOOffspring("GO:0022008",go="BP")  
neuron_death = getGenesForGOOffspring("GO:0070997", go="BP")
neuron_diff = getGenesForGOOffspring("GO:0030182", go="BP")
neuron_bgd = union(union(neurogenesis, neuron_death), neuron_diff)
printFractionalCounts(neuron_bgd, "neuron birth-growth-death")
```

Genes found in both groups?

```{r}
length(intersect(neuron_and_synape, neuron_bgd))
# not sure what other categories to do, or if I should expand BP and MF more. CC seems to show well that we are getting neuron=related expression. I could contrast with another CC?
```


lincRNAs
----------------------

Make dataset.
```{r fig.height=5, fig.width=5}
lincRNA_geneCounts = geneCounts_freeze[which(geneCounts_freeze$biotype == "lincRNA"),]
# get length
linc_lengths = gene_lengths[which(gene_lengths$ensembl_gene_id %in% rownames(lincRNA_geneCounts)),c(1,4)]
range(linc_lengths[,2])
hist(log10(gene_lengths[which(gene_lengths$ensembl_gene_id %in% rownames(lincRNA_geneCounts)),4]), main = "lincRNA gene lengths", xlab = "log10(gene end - gene start)", breaks = 20)
```


Detection by length -- as we add seq depth, are we detecting more longer or shorter lincRNAs?
```{r fig.height=4, fig.width=4}
llincRNAs = which(linc_lengths[,2] > 1e5)
slincRNAs = which(linc_lengths[,2] <= 1e5)
llincRNA_detect = apply(lincRNA_geneCounts[llincRNAs,1:total_samples], MARGIN=2, FUN=countDetected)
slincRNA_detect = apply(lincRNA_geneCounts[slincRNAs,1:total_samples], MARGIN=2, FUN=countDetected)
hist(llincRNA_detect)
hist(slincRNA_detect)
```

Note this includes all samples, not just control samples.

```{r}
plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), llincRNA_detect[1:total_samples]/length(llincRNAs), main = paste("detection of long lincRNAs", "\nn=", length(llincRNAs), sep = ""), xlab = "log10(mapped reads)", ylab = "fraction detected out of total for this dataset", ylim = range(0,1))
lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), llincRNA_detect/length(llincRNAs)), col = "blue", lwd = 2)    
summary(lm(llincRNA_detect/length(llincRNAs) ~ log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)))

plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), slincRNA_detect/length(slincRNAs), main = paste("detection of short-med lincRNAs", "\nn=", length(slincRNAs), sep = ""), xlab = "log10(mapped reads)", ylab = "fraction detected out of total for this dataset", ylim = range(0,1))
lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), slincRNA_detect/length(slincRNAs)), col = "blue", lwd = 2)    
#cor.test(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads),slincRNA_detect/length(slincRNAs))
summary(lm(slincRNA_detect/length(slincRNAs) ~ log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)))
```
