Detection analysis on gene counts for common mind phase I RNAseq data
========================================================
Kristen K Dang for Sage Bionetworks

Load synapse and Ensembl gene counts file.
```{r warning=FALSE}
setwd('~/Computing/commonmind/analysis')
library('rGithubClient')
sageCode = getRepo(repository="kkdang/sage-data-analysis")
sourceRepoFile(sageCode, "rnaseq_analysis_functions.R")
sourceRepoFile(sageCode, "cmc_code/cmc_routines.R")
```


Detection analysis of biotypes
-----------------------------------------------------------

```{r}
head(rownames(geneCounts_freeze))
geneCounts_freeze_temp = geneCounts_freeze
geneCounts_freeze = addBiotype(geneCounts_freeze_temp)
rm(geneCounts_freeze_temp)
totalBiotypeCounts = table(getBM(attributes=c("ensembl_gene_id", "gene_biotype"), mart=Hs)[,2])
```

Plot distribution of number of samples detecting a gene.
```{r}
# Figure 1 and supplemental figure 1
geneCounts_freeze_wdetect = addDetected(geneCounts_freeze[,1:618])
range(geneCounts_freeze_wdetect$detected, na.rm = TRUE)
hist(geneCounts_freeze_wdetect$detected, main = "gene detection -- all genes", xlab="number of samples detecting", ylab = "number of genes", col = "lightcyan2")
pdf(file="fig1_hist_detect.pdf")
hist(geneCounts_freeze_wdetect$detected, main = "gene detection -- all genes", xlab="number of samples detecting", ylab = "number of genes", col = "lightcyan2")
dev.off()
geneCounts_freeze_wdetect$biotype = geneCounts_freeze$biotype
typesToPlot = names(table(geneCounts_freeze_wdetect$biotype))
par(mfrow=c(3,3))
for (i in 1:length(typesToPlot)) {
hist(geneCounts_freeze_wdetect$detected[geneCounts_freeze_wdetect$biotype == typesToPlot[i]], main = typesToPlot[i], xlab = "number samples detecting", col = "lightcyan2", xlim = range(0,700))
}
par(mfrow=c(1,1))
# end figure 1
```


```{r echo=FALSE}
plotDetectionByBiotype=function(inBiotypeCounts, inNormCounts){
  par(mfrow = c(3,3))
  for (i in 1:27) {
    plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inNormCounts[i,], main = rownames(inBiotypeCounts)[i], xlab = "mapped reads", ylab = "fraction detected", ylim = range(0,1))
    lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inNormCounts[i,]), col = "blue", lwd = 2)    
  }
  # interesting categories -- this section for plotting selected results from the above detailed ones
#   par(mfrow = c(2,3))
#   interesting = c(2, 10,16,17,18,21)
#   for (i in 1:length(interesting)) {
#     plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inNormCounts[interesting[i],], main = paste(rownames(inBiotypeCounts)[interesting[i]], "\nn=", EnsemblTotal[interesting[i]], sep = ""), xlab = "mapped reads", ylab = "fraction detected", ylim = range(0,1))
#     lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads), inNormCounts[interesting[i],]), col = "blue", lwd = 2)
#   }
  par(mfrow = c(1,1))
}
```

Plot detection per biotype across varying thresholds for minimum read count. Print linear regression results.
```{r echo=FALSE}
lmResults = matrix(0, nrow = length(totalBiotypeCounts), ncol = 6)
rownames(lmResults) = names(totalBiotypeCounts)
colnames(lmResults) = rep(c("estimate", "pval"), 3)

lmRanges = matrix(0, nrow = length(totalBiotypeCounts), ncol = 6)
rownames(lmRanges) = names(totalBiotypeCounts)
colnames(lmRanges) = rep(c("low", "high"), 3)

lmMeds = matrix(0, nrow = length(totalBiotypeCounts), ncol = 3)
rownames(lmRanges) = names(totalBiotypeCounts)
```

```{r}
j = 1
for (i in 2:4) {
  print(paste('Results for minimum count = ', i, sep = ''))
  biotypeCountsPerSampleFilter = apply(geneCounts_freeze[,1:total_samples], 2, countDetectedByBiotype, biotypes = geneCounts_freeze$biotype, filter=i)
  EnsemblTotal = totalBiotypeCounts[which(names(totalBiotypeCounts) %in% rownames(biotypeCountsPerSampleFilter))]
  normBiotypeCounts = apply(biotypeCountsPerSampleFilter, MARGIN=2, function(x) x/EnsemblTotal)
  plotDetectionByBiotype(inBiotypeCounts=biotypeCountsPerSampleFilter,inNormCounts=normBiotypeCounts)
  coords = match(rownames(normBiotypeCounts), rownames(lmResults))
  lmResults[coords,j:(j+1)] = regressDetectionOnMetadata(inDetectCounts=normBiotypeCounts, inMetaData=log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads))
  lmRanges[coords,j:(j+1)] = t(apply(normBiotypeCounts,MARGIN=1,FUN=range))
  lmMeds[coords,(i-1)] = apply(normBiotypeCounts,MARGIN=1,FUN=median, na.rm = TRUE)
  j = j + 2
}
```


Regression analysis of relationship between detection and depth per biotype.
```{r}
# sets of two columns: first is beta estimate for slope, second is p-value
print(lmResults)
print(lmRanges)
print(lmMeds)
```

Quadrant plot of fraction detected vs regression estimate.
```{r}
#optional figure
toRemove = which(lmResults[,1] == 0)
lmResults_reduced = lmResults[-toRemove,]
lmRanges_reduced = lmRanges[-toRemove,]
lmMeds_reduced = lmMeds[-toRemove,]
plotcols = rainbow(nrow(lmRanges_reduced))
plotcex = log10(totalBiotypeCounts)[-toRemove]
indexToPlot = 1 # values = 1,3,5
altIndex = 1 # values = 1,2,3
#pdf(file="fig_opt_enhanced_scatter.pdf")
plot(lmMeds_reduced[,altIndex],lmResults_reduced[,indexToPlot], xlim = range(0,1), ylim = range(0,1), xlab = "fraction detected", ylab = "genes per depth (slope estimate)", col = plotcols, pch = 16,cex=plotcex)
segments(x0=lmRanges_reduced[,indexToPlot],y0=lmResults_reduced[,indexToPlot],x1=lmRanges_reduced[,(indexToPlot+1)],y1=lmResults_reduced[,indexToPlot], col = plotcols)
legend("topright",legend=rownames(lmRanges_reduced),col=plotcols,pch=16,cex=0.7)
#dev.off()
# end optional figure
```

Now look separately at controls and cases, see if substantially different.
```{r,echo=FALSE}
giantRegressionFunction=function(inCounts,inMeta){
  lmResults = matrix(0, nrow = length(totalBiotypeCounts), ncol = 6)
  rownames(lmResults) = names(totalBiotypeCounts)
  colnames(lmResults) = rep(c("estimate", "pval"), 3)
  
  lmRanges = matrix(0, nrow = length(totalBiotypeCounts), ncol = 6)
  rownames(lmRanges) = names(totalBiotypeCounts)
  colnames(lmRanges) = rep(c("low", "high"), 3)
  
  lmMeds = matrix(0, nrow = length(totalBiotypeCounts), ncol = 3)
  rownames(lmRanges) = names(totalBiotypeCounts)
  j = 1
  for (i in 2:4) {
    print(paste('Results for minimum count = ', i, sep = ''))
    biotypeCountsPerSampleFilter = apply(inCounts, 2, countDetectedByBiotype, biotypes = geneCounts_freeze$biotype, filter=i)
    EnsemblTotal = totalBiotypeCounts[which(names(totalBiotypeCounts) %in% rownames(biotypeCountsPerSampleFilter))]
    normBiotypeCounts = apply(biotypeCountsPerSampleFilter, MARGIN=2, function(x) x/EnsemblTotal)
#    plotDetectionByBiotype(inBiotypeCounts=biotypeCountsPerSampleFilter,inNormCounts=normBiotypeCounts)
    coords = match(rownames(normBiotypeCounts), rownames(lmResults))
    lmResults[coords,j:(j+1)] = regressDetectionOnMetadata(inDetectCounts=normBiotypeCounts, inMetaData=log10(inMeta$DLPFC_RNA_report..Mapped.Reads))
    lmRanges[coords,j:(j+1)] = t(apply(normBiotypeCounts,MARGIN=1,FUN=range))
    lmMeds[coords,(i-1)] = apply(normBiotypeCounts,MARGIN=1,FUN=median, na.rm = TRUE)
    j = j + 2
  }
  toRemove = which(lmResults[,1] == 0)
  lmResults_reduced = lmResults[-toRemove,]
  lmRanges_reduced = lmRanges[-toRemove,]
  lmMeds_reduced = lmMeds[-toRemove,]
  plotcols = rainbow(nrow(lmRanges_reduced))
  plotcex = log10(totalBiotypeCounts)[-toRemove]
  indexToPlot = 1 # values = 1,3,5
  altIndex = 1 # values = 1,2,3
  #pdf(file="fig_opt_enhanced_scatter.pdf")
  plot(lmMeds_reduced[,altIndex],lmResults_reduced[,indexToPlot], xlim = range(0,1), ylim = range(0,1), xlab = "fraction detected", ylab = "genes per depth (slope estimate)", col = plotcols, pch = 16,cex=plotcex)
  segments(x0=lmRanges_reduced[,indexToPlot],y0=lmResults_reduced[,indexToPlot],x1=lmRanges_reduced[,(indexToPlot+1)],y1=lmResults_reduced[,indexToPlot], col = plotcols)
  legend("topright",legend=rownames(lmRanges_reduced),col=plotcols,pch=16,cex=0.7) 
}
```

```{r,cache=TRUE}
table(metadata_freeze$Dx)

geneCounts_control = geneCounts_freeze[,which(metadata_freeze$Dx == "Control")]
metadata_control = metadata_freeze[which(metadata_freeze$Dx == "Control"),]
giantRegressionFunction(inCounts=geneCounts_control,inMeta=metadata_control)

geneCounts_case = geneCounts_freeze[,which(metadata_freeze$Dx %in% c("SCZ", "BP"))]
metadata_case = metadata_freeze[which(metadata_freeze$Dx %in% c("SCZ", "BP")),]
giantRegressionFunction(inCounts=geneCounts_case,inMeta=metadata_case)

```




Look at fractional allocation of reads to biotypes across depths. Maybe look in controls only?
```{r}
biotypeCounts = apply(geneCounts_freeze[,1:618],MARGIN=2,function(x) tapply(x, geneCounts_freeze$biotype, FUN=sum))
biotypeProportions = apply(biotypeCounts,MARGIN=2,function(x) x/sum(x))
hist(biotypeProportions[18,])

# Looks like fraction protein-coding decreases with depth, but nothing else changes much
rownames(biotypeProportions)
regressProportions=function(proportions){
  x = summary(lm(proportions~log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)+metadata_freeze$Dx))
  return(c(x$coefficients[2,1],x$coefficients[2,4]))
  }
effectOfDepthOnFractions = apply(biotypeProportions,MARGIN=1,regressProportions)
```

Specifically, what fraction of reads are pseudogenes with low detection? These presumably are mis-alignments that take reads away from correct matches.
```{r}
lowDetectPseudo = intersect(which(geneCounts_freeze_wdetect$biotype == "pseudogene"), which(geneCounts_freeze_wdetect$detected < 550))
lowDetectPseudoSums = apply(geneCounts_freeze[lowDetectPseudo,1:618],MARGIN=2,sum)
hist(lowDetectPseudoSums, main = "number of reads attributable\nto pseudogenes with low detection", xlab = "pseudocount sum per sample", col = "lightcyan2")
hist(lowDetectPseudoSums/colSums(geneCounts_freeze[,1:618]), main = "fraction of reads attributable to\npseudogenes with low detection", xlab = "pseudocount sum / total sum", col = "lightcyan2")
```

Detection analysis of expression levels, all biotypes together.
---------------------------------------------------------------

```{r fig.height=4, fig.width=4}
gene_lengths = getGeneLengths()
hist(log10(gene_lengths$length), main='Distribution of gene lengths by gene start stop')
```

Looking at control samples only.

```{r echo=c(2:6), fig.height=6, fig.width=6}
library('edgeR')
table(metadata_freeze$Dx)
data.dge = DGEList(counts=geneCounts_freeze[,1:total_samples],group=factor(metadata_freeze$Dx), remove.zeros=TRUE)
data.dge = calcNormFactors(data.dge)
control.dge = data.dge[,which(data.dge$samples$group == "Control")]
control_gene_means = rowMeans(rpkm(control.dge, gene.length=gene_lengths$length[match(rownames(getCounts(control.dge)), gene_lengths$ensembl_gene_id)], log=TRUE,normalized.lib.sizes=TRUE))
```

```{r fig.height=6, fig.width=6}
hist(rpkm(control.dge, gene.length=gene_lengths$length[match(rownames(getCounts(control.dge)), gene_lengths$ensembl_gene_id)],log=TRUE,normalized.lib.sizes=TRUE), main = "control samples, all genes", col = "thistle", xlab = "RPKM")
hist(control_gene_means, main = "gene averages across samples\ncontrol samples only", col = "thistle", breaks = 50)
```

Separate into high/med/low expression datasets.

```{r}
range(control_gene_means, na.rm = TRUE)
lowExpress = which(control_gene_means < -7)
highExpress = which(control_gene_means > -2)
medExpress_temp = which(control_gene_means >= -7)
medExpress = setdiff(medExpress_temp, highExpress)
sum(length(lowExpress), length(highExpress), length(medExpress))
rm(medExpress_temp)

control_high.dge = control.dge[highExpress,]
control_med.dge = control.dge[medExpress,]
control_low.dge = control.dge[lowExpress,]

control_high_detect = apply(getCounts(control_high.dge), MARGIN=2, FUN=countDetected)
control_med_detect = apply(getCounts(control_med.dge), MARGIN=2, FUN=countDetected)
control_low_detect = apply(getCounts(control_low.dge), MARGIN=2, FUN=countDetected)
```

Plot relationships between fraction detected and mapped reads by expression level.

```{r}
#low
plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_low_detect/length(lowExpress), main = paste("detection of low-expression genes", "\nn=", length(lowExpress), sep = ""), xlab = "log10(mapped reads)", ylab = "fraction detected", ylim = range(0,1))
lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_low_detect/length(lowExpress)), col = "blue", lwd = 2)    
cor.test(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_low_detect)

# medium
plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_med_detect/length(medExpress), main = paste("detection of med-expression genes", "\nn=", length(medExpress), sep = ""), xlab = "log10(mapped reads)", ylab = "fraction detected", ylim = range(0,1))
lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_med_detect/length(medExpress)), col = "blue", lwd = 2)    
cor.test(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_med_detect)


# high
plot(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_high_detect/length(highExpress), main = paste("detection of high-expression genes", "\nn=", length(highExpress), sep = ""), xlab = "log10(mapped reads)", ylab = "fraction detected", ylim = range(0,1))
lines(lowess(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_high_detect/length(highExpress)), col = "blue", lwd = 2)    
cor.test(log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)[which(metadata_freeze$Dx == "Control")], control_high_detect)
```


What genes are expressed at hi/med/low levels? Do they look like neurons?
--------------------------------------

```{r, echo=FALSE}
## trying to look at categories in each group
# pcGenes = getByBiotype()
# highExpress_ENSGpc = intersect(rownames(getCounts(control_high.dge)), pcGenes[,1])
# medExpress_ENSGpc = intersect(rownames(getCounts(control_med.dge)), pcGenes[,1])
# lowExpress_ENSGpc = intersect(rownames(getCounts(control_low.dge)), pcGenes[,1])
# write.csv(lowExpress_ENSGpc, file = "lowexpress_ENSGpc", quote = FALSE, row.names = FALSE, col.names = FALSE)
# write.csv(highExpress_ENSGpc, file = "highexpress_ENSGpc", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

Getting neuro-relevant genes from GO, then looking at expression level of those genes in our samples.

```{r,echo=FALSE}
printFractionalCounts=function(categoryENSGvec, categoryName){
  print(paste("Number of genes in category", length(categoryENSGvec), sep = " "))
  print(paste("High expression genes as fraction of all genes:", length(highExpress) / nrow(control.dge), sep = " "))
  print(paste("Fraction of", categoryName, "genes found in high expression group:", length(intersect(rownames(getCounts(control_high.dge)), categoryENSGvec))/length(categoryENSGvec), sep = " "))
  
  print(paste("Medium expression genes as fraction of all genes:", length(medExpress) / nrow(control.dge), sep = " "))
  print(paste("Fraction of", categoryName, "genes found in med expression group:", length(intersect(rownames(getCounts(control_med.dge)), categoryENSGvec))/length(categoryENSGvec), sep = " "))
  
  print(paste("Low expression genes as fraction of all genes:", length(lowExpress) / nrow(control.dge), sep = " "))
  print(paste("Fraction of", categoryName, "genes found in low expression group:", length(intersect(rownames(getCounts(control_low.dge)), categoryENSGvec))/length(categoryENSGvec), sep = " "))  
}
```

Genes in neuron-related cellular component GO
```{r cache=TRUE, warning=FALSE}
neuron_part = getGenesForGOOffspring("GO:0097458")
synapse = getGenesForGOOffspring("GO:0045202")
vesicle = getGenesForGOOffspring("GO:1990008") # nothing found
neuron_and_synape = union(neuron_part, synapse)
printFractionalCounts(neuron_and_synape, "neuron-synapse")
```

Genes in neuron birth-growth-death biological process GO.

```{r cache=TRUE}
neurogenesis = getGenesForGOOffspring("GO:0022008",go="BP")  
neuron_death = getGenesForGOOffspring("GO:0070997", go="BP")
neuron_diff = getGenesForGOOffspring("GO:0030182", go="BP")
neuron_bgd = union(union(neurogenesis, neuron_death), neuron_diff)
printFractionalCounts(neuron_bgd, "neuron birth-growth-death")
```

Genes found in both groups?

```{r}
length(intersect(neuron_and_synape, neuron_bgd))
# not sure what other categories to do, or if I should expand BP and MF more. CC seems to show well that we are getting neuron=related expression. I could contrast with another CC?
```

