Detection analysis on gene counts by RIN
========================================================
Kristen K Dang for Sage Bionetworks
April 3, 2014

Which genes are over/undersampled with low RIN?
Start with control samples only.

Load synapse and Ensembl gene counts file.
```{r warning=FALSE}
setwd('~/Computing/commonmind/analysis')
library('rGithubClient')
sageCode = getRepo(repository="kkdang/sage-data-analysis")
sourceRepoFile(sageCode, "rnaseq_analysis_functions.R")
sourceRepoFile(sageCode, "cmc_code/cmc_routines.R")
```

```{r}
table(metadata_freeze$Dx)
geneCounts_control = geneCounts_freeze[,which(metadata_freeze$Dx == "Control")]
metadata_control = metadata_freeze[which(metadata_freeze$Dx == "Control"),]
```


Plot distribution of RIN
```{r}
hist(metadata_control$DLPFC_RNA_isolation..RIN, breaks = 50, col = "peru", main = "sample quality\ncontrol samples only")
hist(metadata$DLPFC_RNA_isolation..RIN, breaks = 50, col = "peru", main="sample quality\nall samples")
lowRIN = which(metadata_control$DLPFC_RNA_isolation..RIN < 7)
length(lowRIN)
lowRIN_counts = geneCounts_control[,lowRIN]
highRIN = which(metadata_control$DLPFC_RNA_isolation..RIN > 8)
length(highRIN)
highRIN_counts = geneCounts_control[,highRIN]
```

What genes have largest changes in detection across samples when comparing samples with low RIN to samples with high RIN? Is composition of biotypes changing with RIN? 
```{r}
lowRIN_detect = apply(lowRIN_counts,1,countDetected)
highRIN_detect = apply(highRIN_counts,1,countDetected)
boxplot(lowRIN_detect/ncol(lowRIN_counts), highRIN_detect/ncol(highRIN_counts),names=c("low RIN", "high RIN"), ylab = "fraction of genes detected across samples")
hist(lowRIN_detect/ncol(lowRIN_counts) - highRIN_detect/ncol(highRIN_counts), main = "change in detection between\nhigh-low qual samples")
```

Genes with largest decrease of detection with higher RIN
```{r}
dDetect = rownames(geneCounts_control)[which((lowRIN_detect/ncol(lowRIN_counts) - highRIN_detect/ncol(highRIN_counts)) > 0.5)]
dDetect_tmp = data.frame(rep(NA, length(dDetect)))
rownames(dDetect_tmp) = dDetect
table(addBiotype(dDetect_tmp)$biotype)
```

Genes with largest increase of detection with higher RIN
```{r}
iDetect = rownames(geneCounts_control)[which((lowRIN_detect/ncol(lowRIN_counts) - highRIN_detect/ncol(highRIN_counts)) < -0.5)]
iDetect_tmp = data.frame(rep(NA, length(iDetect)))
rownames(iDetect_tmp) = iDetect
table(addBiotype(iDetect_tmp)$biotype)
```
Looks like higher RIN finds fewer pseudogenes. This suggests reads are aligning erroneously to the pseudogene rather than the gene, presumably due to poor quality libraries.



Which genes have significant association with detection by RIN? Per-gene logistic regression on detection vs RIN+depth. 
```{r,cache=TRUE,warning=FALSE}
# This function looks at only control genes
detectFilter=function(count,filter=2){
  if (count > filter) { return(1)}
  else { return(0) }
}
runLogistic=function(gCounts){
  F = sapply(gCounts,FUN=detectFilter, simplify=TRUE)
  if (sum(F) < length(gCounts)){
    fit = glm(F~metadata_control$DLPFC_RNA_isolation..RIN+log10(metadata_control$DLPFC_RNA_report..Mapped.Reads),family=binomial())
    temp = summary(fit)
    return(c(temp$coefficients[2,4], temp$coefficients[2,1]) )
  }
  else{return(c(NA, NA))}
}

# RINeffectByGene = apply(X=geneCounts_control,MARGIN=1,FUN=runLogistic)
# hist(RINeffectByGene[1,], col = "lightcyan2")
# hist(RINeffectByGene[1,], col = "lightcyan2", xlim = range(0,0.01), breaks = 800)
# sigGenes = which(RINeffectByGene[1,] < .001)
# length(sigGenes)
```


Alternate version, using all samples but controlling for Dx.
```{r,cache=TRUE,warning=FALSE}
runLogistic_wDx=function(gCounts){
  F = sapply(gCounts,FUN=detectFilter, simplify=TRUE)
  if (sum(F) < length(gCounts)){
    fit = glm(F~metadata_freeze$DLPFC_RNA_isolation..RIN+log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)+metadata_freeze$Dx,family=binomial())
    temp = summary(fit)
    return(c(temp$coefficients[2,4], temp$coefficients[2,1]) )
  }
  else{return(c(NA, NA))}
}

apply(X=geneCounts_freeze[2,],MARGIN=1,FUN=runLogistic_wDx)
RINeffectByGene_wDx = apply(X=geneCounts_freeze,MARGIN=1,FUN=runLogistic_wDx)
hist(RINeffectByGene_wDx[1,], col = "lightcyan2", main = "pvalues of detection-RIN relationship")
sigGenes = which(RINeffectByGene_wDx[1,] < .001)
length(sigGenes)
```


Which genes have a significant increase in detection by RIN?
```{r}
hist(RINeffectByGene_wDx[2,sigGenes], col = "lightcyan2", main = "coefficient estimates for RIN")
sigGenesIncrease = which((RINeffectByGene_wDx[1,] < .001) & (RINeffectByGene_wDx[2,] >0))
iRINsig = data.frame(t(RINeffectByGene_wDx[,sigGenesIncrease]))
rownames(iRINsig) = names(sigGenesIncrease)
iRINsig_wBiot = addBiotype(iRINsig)
table(iRINsig_wBiot$biotype)
#write.table(rownames(iRINsig_wBiot)[which(iRINsig_wBiot$biotype == "protein_coding")], row.names = F, quote = F)

# background detection file for use with CPDB gene over-representation analysis
geneSums = rowSums(geneCounts_control)
write.table(names(which(geneSums > 0)),file="detected_control_min1.txt", row.names=F,quote=F,sep='\n',col.names=F)
```

Which genes have a significant decrease in detection by RIN?
```{r}
sigGenesDecrease = which((RINeffectByGene_wDx[1,] < .001) & (RINeffectByGene_wDx[2,] < 0))
dRINsig = data.frame(RINeffectByGene_wDx[sigGenesDecrease])
rownames(dRINsig) = names(sigGenesDecrease)
table(addBiotype(dRINsig)$biotype)
```


Boxplot RIN values vs detection status for a few genes
```{r}
temp_data = data.frame(sapply(geneCounts_control[which(rownames(geneCounts_control) == rownames(dRINsig)[1]),],FUN=detectFilter, simplify=TRUE))
colnames(temp_data) = c('detected')
temp_data$RIN = metadata_control$DLPFC_RNA_isolation..RIN
boxplot(temp_data$RIN~temp_data$detected,names=c("not", "detected"), col = "thistle1", ylab = 'RIN', main = 'sample RIN values grouped by\ndetection status for that sample',varwidth=T)
```


Setting up another logistic regression, where expression level of parent gene is taken into account. First get information on 'parent' genes of pseudogoenes.
```{r}
#psiSetFile = synGet('syn2400043')
dropSuffix=function(ID){
  unlist(strsplit(as.character(ID),split="[.]"))[1]
}
ensgt_matrix = read.delim(file='~/Computing/commonmind/data/ensg-enst-parse.txt',header=FALSE)
psiFile = synGet('syn2400033')
psiData  = read.delim(psiFile@filePath,skip=20,header=TRUE)
psiData_parsedID = sapply(psiData$Pseudogene_id,dropSuffix)
temp = psiData$Parent_gene[match(ensgt_matrix[,2], psiData_parsedID)]
ensgt_matrix$parent = sapply(temp,dropSuffix)
rm(temp)
```


Reduce the input to a set of pseudogenes for which there is a parent gene with counts data in our dataset.
```{r}
geneCounts_freeze_temp = geneCounts_freeze
geneCounts_freeze = addBiotype(geneCounts_freeze_temp)
rm(geneCounts_freeze_temp)
pseudoCounts = geneCounts_freeze[which(geneCounts_freeze$biotype == "pseudogene"),1:(ncol(geneCounts_freeze)-1)]

# get positions of pseudocounts genes in ensgt_matrix
z = match(rownames(pseudoCounts), ensgt_matrix[,1])
# remove those that don't have an entry in ensgt_matrix
pseudoCounts_tmp = pseudoCounts[-which(is.na(z)),]
# get positions of pseudocounts genes in ensgt_matrix again
z = match(rownames(pseudoCounts_tmp), ensgt_matrix[,1])
# reduce ensgt_matrix to the ones with a match in pseudoCounts, in order of pseudo
ensgt_matrix_short = ensgt_matrix[z,]
head(rownames(pseudoCounts_tmp))
head(ensgt_matrix_short)

# remove pseudoCounts entries that don't have a parent gene in ensgt_matrix_short
pseudoParentIndicies_tmp = match((ensgt_matrix_short$parent), rownames(geneCounts_freeze)) 
pseudoCounts_tmp2 = pseudoCounts_tmp[-which(is.na(pseudoParentIndicies_tmp)),]
dim(pseudoCounts_tmp2)
# get positions of pseudocounts genes in ensgt_matrix again
z = match(rownames(pseudoCounts_tmp2), ensgt_matrix_short[,1])
ensgt_matrix_short2 = ensgt_matrix_short[z,]
head(ensgt_matrix_short2)
dim(ensgt_matrix_short2)
head(rownames(pseudoCounts_tmp2))

# get parentCount matrix based on entries now present in ensgt_matrix_short2
pseudoParentIndicies = match((ensgt_matrix_short2$parent), rownames(geneCounts_freeze)) 
pseudoParentCounts = geneCounts_freeze[pseudoParentIndicies,1:(ncol(geneCounts_freeze)-1)]
```

Run the regression and plot results. 
```{r,warning=FALSE}
runLogistic_wDx_andParent=function(gCounts,parentCounts){
  F = sapply(gCounts,FUN=detectFilter, simplify=TRUE)
  if ((sum(F) < length(gCounts))&(sum(F) > 0)){
    parentExp = log10(parentCounts/metadata_freeze$DLPFC_RNA_report..Mapped.Reads*1e6)
    parentExp[which(parentExp %in% c("Inf", "-Inf"))] = NA
    fit = try(glm(F~metadata_freeze$DLPFC_RNA_isolation..RIN+log10(metadata_freeze$DLPFC_RNA_report..Mapped.Reads)+metadata_freeze$Dx+parentExp,family=binomial()),silent=T)
    if(length(fit) > 1) {
      temp = summary(fit)
      if (nrow(temp$coefficients) < 6) { return(rep(NA,6)) }
      return(c(temp$coefficients[2,4], temp$coefficients[2,1], temp$coefficients[3,4], temp$coefficients[3,1],temp$coefficients[6,4], temp$coefficients[6,1] ))
    }
    else{return(rep(NA, 6))}
  }
  else{return(rep(NA, 6))}
}

tpseudocounts2 = data.frame(t(pseudoCounts_tmp2))
tparentcounts = data.frame(t(pseudoParentCounts))
mapply(runLogistic_wDx_andParent, gCounts=tpseudocounts2[,348:349], parentCounts=tparentcounts[,348:349])
RINeffectByGene_wDx_pseudo = mapply(runLogistic_wDx_andParent, gCounts=tpseudocounts2, parentCounts=tparentcounts)
hist(RINeffectByGene_wDx_pseudo[1,], col = "lightcyan2", main = "pval RIN")
hist(RINeffectByGene_wDx_pseudo[2,], col = "lightcyan2", main = "est RIN", xlim = range(-100,100), breaks = 2000)
hist(RINeffectByGene_wDx_pseudo[5,], col = "lightcyan2", main = "pval parent")
hist(RINeffectByGene_wDx_pseudo[6,], col = "lightcyan2", main = "est parent", xlim = range(-10000,10000), breaks = 200)

# Test the reduced pseudogenes only against the linear model
RINeffectByGene_wDx_pseudofinal = apply(X=pseudoCounts_final,MARGIN=1,FUN=runLogistic_wDx_andParent)
hist(RINeffectByGene_wDx_pseudofinal[1,], col = "lightcyan2")
```

What fraction of pseudogenes tested do not change detection with respect to RIN or parent gene expression? This is an upper bound on valid fraction pseudogenes (ie true observations).
```{r}
length(which(RINeffectByGene_wDx_pseudo[1,] < 0.01)) # RIN
length(which(RINeffectByGene_wDx_pseudo[3,] < 0.01)) # library size
length(which(RINeffectByGene_wDx_pseudo[5,] < 0.01)) # parent expression

parentAndRIN = union(which(RINeffectByGene_wDx_pseudo[1,] < 0.01), which(RINeffectByGene_wDx_pseudo[5,] < 0.01)) # either RIN or parent
length(parentAndRIN)
nonZeroResult = ncol(RINeffectByGene_wDx_pseudo) - length(which(is.na(RINeffectByGene_wDx_pseudo[1,])==TRUE)) # number of genes for which we were able to test (ie not an NA result)
nonZeroResult
fractionPotentiallyValid = 1-(length(parentAndRIN)/nonZeroResult)
fractionPotentiallyValid
```