# KKD for Sage Bionetworks
# Nov. 12, 2014

```{r message=FALSE,warning=FALSE}
library('synapseClient')
library('edgeR')
synapseLogin()
library('rGithubClient')
require(goProfiles)
sageCode = getRepo(repository="kkdang/sage-data-analysis")
sourceRepoFile(sageCode, "rnaseq_analysis_functions.R")
setwd('/Users/kristen/Computing/cranio/')
```


Alignment results
```{r cache=TRUE}
alignEntity = synGet('syn2820392')
align = read.csv(getFileLocation(alignEntity))
colnames(align)
hist(align$CODING_BASES/250)
hist(align$CORRECT_STRAND_READS/2)
```


Experimental variables
```{r cache=TRUE}
clinicalTable = synTableQuery('SELECT * FROM syn2823605')
```

Counting results summary
```{r cache=TRUE}
# Using data as Table in Synapse
metricsTable = synTableQuery('SELECT * FROM syn3105973')
metrics = metricsTable@values
rownames(metrics) = metrics[,1]
metrics = data.matrix(metrics[,2:10])
pie(colMeans(metrics))
hist(rowSums(metrics), main = "distribution across samples of total counts\n(read pairs)", col = "powderblue")
plotFCfractionalResults(metrics)

# Look at characteristics of low-count samples.
# Which samples have fewer than 20 million assigned read pairs?
lowReads = names(which(metrics[,1] < 20e6))
# Are they cases or controls? 
table(clinicalTable@values$"Sample Type"[match(lowReads,as.character(clinicalTable@values$SeqSampleName))])
table(clinicalTable@values$"Sample Type")

```

Sample-level expression values
```{r cache=TRUE}
dataEntity = synGet('syn2820309') # estimated Reads
library('R.utils')
gunzip(getFileLocation(dataEntity),overwrite=TRUE)
x = unlist(strsplit(getFileLocation(dataEntity), split = '.gz'))
detach("package:R.utils", unload=TRUE)
geneCounts = read.csv(x, row.names = 1)

controlSampleIDs = clinicalTable@values$SeqSampleName[which(clinicalTable@values$"Sample Type" == "Control")]
casesStatus = rep("case", ncol(geneCounts))
casesStatus[which(colnames(geneCounts) %in% paste("X", controlSampleIDs, sep=""))] = "control"

data.dge = DGEList(counts=geneCounts,group=factor(casesStatus),remove.zeros=TRUE)
data_palx.dge = filterByFractionPresent(data.dge,fraction=0.05)
data_palx.dge = calcNormFactors(data_palx.dge)
control.dge = data_palx.dge[,which(data_palx.dge$samples$group == "control")]
cases.dge = data_palx.dge[,which(data_palx.dge$samples$group == "case")]


# Library sizes
hist(data_palx.dge$samples$lib.size)

## kernel density plots
densityPlot(control.dge, mainLabel="control",ub=0.2)
densityPlot(cases.dge, mainLabel="cases",ub=0.2)

## Boxplots
boxplot(cpm(data_palx.dge,normalized.lib.sizes=TRUE,log=TRUE),outline=FALSE, las = 2, main = "distribution of expression values", xlab = "sample", ylab = "log2(counts per million)",names=NA)
```


Quick look at group level differences
```{r cache=TRUE}
control.ave = aveLogCPM(control.dge)
cases.ave = aveLogCPM(cases.dge)
boxplot(control.ave, main = "ave gene expression, controls")
boxplot(cases.ave, main = "ave gene expression, cases")

plotColors = c("red", "blue")
plot(density(cases.ave), col = "red", lwd = 2, main = "per-sample ave density plot")
lines(density(control.ave), col = "blue", lwd = 2)
legend("topright", legend=c("cases", "control"), col=plotColors,pch=16)
```


What genes have high expression?
```{r cache=TRUE}
highCases = which(cases.ave > 10)
highControls = which(control.ave > 10)
highBoth = intersect(highCases, highControls)
highExpGenes = rownames(data_palx.dge)[highCases]
highExpGenes_vals = rep(0, length(highExpGenes))
names(highExpGenes_vals) = highExpGenes
#write.table(highExpGenes,file="list_highExpGenes.txt",quote=FALSE,row.names=FALSE,col.names=FALSE)

#library('topGO')
#sampleGOdata <- new("topGOdata",ontology = "BP",allGenes = highExpGenes_vals, geneSel = function(p) {p < 0.01}, annot = annFUN.org, mapping = "org.Hs.eg.db", ID = "Ensembl")
#resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
#allRes <- GenTable(sampleGOdata, classicFisher = resultFisher,ranksOf = "classicFisher", topNodes = 100)

```


Profile detected genes
```{r cache=TRUE}
#Make three versions of the dataset where minimum reads required for detection is varied from 2:4.
rawCounts = getCounts(data_palx.dge)

detected_control = apply(rawCounts, MARGIN=1, FUN=countDetected)
hist(detected_control, col = "papayawhip")

detected3_control = apply(rawCounts, MARGIN=1, FUN=countDetected, filter=3)
hist(detected3_control, col = "papayawhip")

detected4_control = apply(rawCounts, MARGIN=1, FUN=countDetected, filter=4)
hist(detected4_control, col = "papayawhip")

# Assign Entrez ID to genes and remove genes with no Entrez ID
temp = addEntrez(inCounts=as.data.frame(detected4_control))
toRemove = which(is.na(temp$entrez))
detect_entrez = temp[-toRemove,]
rm(temp)

detectFilter_entrez = unique(detect_entrez$entrez[which(detect_entrez$detected4_control > 0)])
```



Functional profile of high expression genes
```{r cache=TRUE}
temp = addEntrez(inCounts=as.data.frame(highExpGenes_vals))
toRemove = which(is.na(temp$entrez))


highexp.mf5.profiles = basicProfile(genelist=temp$entrez,onto="MF",level=5,orgPackage="org.Hs.eg.db", empty.cats=FALSE)
#save(highexp.mf5.profiles,file="highexp_mf5_profiles.bz2",compress="bzip2")
#load("highexp_mf5_profiles.bz2")
#print(highexp.mf5.profiles, percentage=TRUE)
#hist(highexp.mf5.profiles[[1]]$Frequency, breaks = 50)
z = which(highexp.mf5.profiles[[1]]$Frequency < 2)
highexp.mf5.profiles.reduced = highexp.mf5.profiles
highexp.mf5.profiles.reduced[[1]] = highexp.mf5.profiles.reduced[[1]][-z,]
rm(z)
plotProfiles(highexp.mf5.profiles.reduced, "Major functional profiles\nhigh expression genes",percentage = F, multiplePlots = F,labelWidth=28)

```