---
title: "variance_analysis.Rmd"
author: "Kristen Dang"
date: "December 19, 2014"
output: html_document
---

```{r message=FALSE,warning=FALSE}
library('FactoMineR')
library('synapseClient')
synapseLogin()

library('rGithubClient')
sageCode = getRepo(repository="kkdang/sage-data-analysis")
sourceRepoFile(sageCode, "rnaseq_analysis_functions.R")
source('~/Computing/sage-data-analysis/scri_cran/cranio_common_routines.R')
setwd('/Users/kristen/Computing/cranio/')
cutoff = 0.01
```


Function to make PC datasets with different filters 
```{r}
proteinCoding_transcripts = getByBiotype(biotype = 'protein_coding',gene = FALSE)
proteinCoding_genes = getByBiotype(biotype = 'protein_coding',gene = TRUE)
FM_PCA=function(in.dge){
  result.prc = PCA(X = t(cpm(in.dge,normalized.lib.sizes=TRUE)),scale.unit = TRUE,ncp = 5,graph = FALSE)
#  barplot(result.prc$eig$"percentage of variance", main = "percentage of variance per pc")
  barplot(result.prc$eig$"percentage of variance"[1:20], ylab = "% variance") 
  return(result.prc)
}

prcomp_PCA=function(in.dge){
  prc = prcomp(t(cpm(in.dge,normalized.lib.sizes=TRUE)),center=TRUE,scale=TRUE,retx = TRUE)
  barplot((prc$sdev^2/sum(prc$sdev^2))[1:20], ylab = "fraction variance") 
  return(prc)
}

generate_PCA_datasets=function(in.dge,pcNames,factoMine=FALSE){
  
  #### Make several filtered datasets ####

  ## All data, PALO
  palo.dge = calcNormFactors(in.dge)
  
  ## PALX=5%
  palx5.dge = filterByFractionPresent(in.dge,fraction=0.05,minCount=1)
  palx5.dge = calcNormFactors(palx5.dge)
  
  ## PALX=30%
  palx30.dge = filterByFractionPresent(in.dge,fraction=0.30,minCount=1)
  palx30.dge = calcNormFactors(palx30.dge)
  
  ## Protein-coding genes, PALO
  b = as.list(rownames(in.dge))
  strippedNames = sapply(b,function(x){unlist(strsplit(as.character(x), split = "[.]"))[1]})  
  pc.dge = in.dge[which(strippedNames %in% pcNames[,1]),]
  pc_palo.dge = DGEList(counts = getCounts(pc.dge),group = pc.dge$samples$group)
  pc_palo.dge = calcNormFactors(pc_palo.dge)
  
  ## Protein-coding genes, PALX=5%
  pc_palx5.dge = filterByFractionPresent(pc_palo.dge,fraction=0.05,minCount=1)
  pc_palx5.dge = calcNormFactors(pc_palx5.dge)
  
  ## Protein-coding genes, PALX=20%
  pc_palx20.dge = filterByFractionPresent(pc_palo.dge,fraction=0.2,minCount=1)
  pc_palx20.dge = calcNormFactors(pc_palx20.dge)
  
  
  ### PCA via SVD on standardized,centered cpm expression values ####
  if (factoMine) {
    op = par(mfrow = c(2,3))
    FM_PCA(palo.dge)
    FM_PCA(palx5.dge)
    FM_PCA(palx30.dge)
    FM_PCA(pc_palo.dge)
    FM_PCA(pc_palx5.dge)
    FM_PCA(pc_palx20.dge)
    par(op)  
  }
  
  else {
    op = par(mfrow = c(2,3))
    prcomp_PCA(palo.dge)
    prcomp_PCA(palx5.dge)
    prcomp_PCA(palx30.dge)
    prcomp_PCA(pc_palo.dge)
    prcomp_PCA(pc_palx5.dge)
    prcomp_PCA(pc_palx20.dge)
    par(op)
  }
  return(pc_palo.dge)
}
```


Functions to calculate correlation with PCs and clinical/exp factors
```{r}
pcaCorrelation=function(x,y){cor.test(x,y)$p.value }
pcaCorrelationVal=function(x,y){cor.test(x,y)$estimate }
pcaLM=function(x,y){summary(lm(x ~ y))$coefficients[2,4]}

calc_plot_PC_corr=function(in.pca,inClinical=clinicalDataR){
  # Using correlation against eigenvectors from all PALO data
  clinicalPCAcorrelations = matrix(NA, nrow = nrow(in.pca), ncol = ncol(inClinical)-1)
  colnames(clinicalPCAcorrelations) = rep("default", ncol(clinicalPCAcorrelations))
  
  k = 1
  # correlate numeric factors on log scale
  toRun = c(5,10,15)
  for (j in k:length(toRun)) {
    clinicalPCAcorrelations[,j] = apply(in.pca,MARGIN=2,pcaCorrelation,y=log(inClinical[,toRun[j]]))
    colnames(clinicalPCAcorrelations)[j] = colnames(inClinical)[toRun[j]]
    }
  k = k + length(toRun) -1
  
  # correlate one numeric factor on non-log since it is already normal
  toRun = c(11)
  for (j in 1:length(toRun)) {
    clinicalPCAcorrelations[,k+j] = apply(in.pca,MARGIN=2,pcaCorrelation,y=inClinical[,toRun[j]])
    colnames(clinicalPCAcorrelations)[k+j] = colnames(inClinical)[toRun[j]]
    }
  k = k + length(toRun) 
  
  
  # correlate against all numeric dates
  toRun = c(8,9,12,13,14)
  for (j in 1:length(toRun)) {
    clinicalPCAcorrelations[,k+j] = apply(in.pca,MARGIN=2,pcaCorrelation,y=as.numeric(inClinical[,toRun[j]]))
    colnames(clinicalPCAcorrelations)[k+j] = colnames(inClinical)[toRun[j]]
    }
  k = k + length(toRun) 
  
  # regress against categorical factors
  toRun = c(6,7,16,17)
  for (j in 1:length(toRun)) {
    clinicalPCAcorrelations[,k+j] = apply(in.pca,MARGIN=2,pcaLM,y=inClinical[,toRun[j]])
    colnames(clinicalPCAcorrelations)[k+j] = colnames(inClinical)[toRun[j]]
    }
  
  op = par(mai = c(3,1,1,1))
  boxplot(clinicalPCAcorrelations[,1:13], las = 2)
  par(op)
  boxplot(t(clinicalPCAcorrelations[1:20,1:13]), las = 2, main = "Correlation of each PCs vs each factor", ylab = "raw pvalue")
  dotchart(t(clinicalPCAcorrelations[1:2,1:13]), xlab = "pvalue", main = "Correlation with PC1 and PC2",bg="tomato",lcolor="tomato",cex=1.1)
  dotchart(t(clinicalPCAcorrelations[1:2,1:13]), xlab = "pvalue", main = "Correlation with PC1 and PC2", xlim = range(0,0.1),bg="tomato",lcolor="tomato",cex=1.1)
  abline(v=0.05, lty = 2)
  abline(v = 0.01, lty = 3)
  #apply(clinicalPCAcorrelations[1:20,],MARGIN=1,stripchart, las = 2)
}
```


Function to correlate with alignment metrics
```{r}

calc_plot_alignPC_corr=function(in.prc,inAlign){
  # Using correlation against eigenvectors
  alignPCAcorrPval = matrix(NA, nrow = nrow(in.prc), ncol = ncol(inAlign))
  colnames(alignPCAcorrPval) = rep("default", ncol(alignPCAcorrPval))
  alignPCAcorrelations = matrix(NA, nrow = nrow(in.prc), ncol = ncol(inAlign))
  colnames(alignPCAcorrelations) = rep("default", ncol(alignPCAcorrelations))
  
  # correlate percentages and bias fractions
  for (j in 1:ncol(inAlign)) {
    alignPCAcorrPval[,j] = apply(in.prc,MARGIN=2,pcaCorrelation,y=inAlign[,j])
    colnames(alignPCAcorrPval)[j] = colnames(inAlign)[j]
    alignPCAcorrelations[,j] = apply(in.prc,MARGIN=2,pcaCorrelationVal,y=inAlign[,j])
    colnames(alignPCAcorrelations)[j] = colnames(inAlign)[j]
    }
  
  # Pvalues for the correlation test
  op = par(mai = c(3,1,1,1))
  boxplot(alignPCAcorrPval[,1:ncol(alignPCAcorrPval)], las = 2)
  par(op)
  boxplot(t(alignPCAcorrPval[1:15,1:ncol(alignPCAcorrPval)]), las = 2, main = "Correlation of each PCs vs each factor", ylab = "raw pvalue")
  dotchart(t(alignPCAcorrPval[1:2,1:ncol(alignPCAcorrPval)]), xlab = "pvalue", main = "Correlation with PC1 and PC2",bg="tomato",lcolor="tomato")
  dotchart(t(alignPCAcorrPval[1:2,1:ncol(alignPCAcorrPval)]), xlab = "pvalue", main = "Correlation with PC1 and PC2", xlim = range(0,0.1),bg="tomato",lcolor="tomato")
  abline(v=0.05, lty = 2)
  abline(v = 0.01, lty = 3)
  
  # Correlation values
  op = par(mai = c(3,1,1,1))
  boxplot(alignPCAcorrelations[,1:ncol(alignPCAcorrelations)], las = 2)
  par(op)
  boxplot(t(alignPCAcorrelations[1:15,1:ncol(alignPCAcorrelations)]), las = 2, main = "Correlation of each PCs vs each factor", ylab = "correlation estimate")
  dotchart(t(alignPCAcorrelations[1:2,1:ncol(alignPCAcorrelations)]), xlab = "correlation estimate", main = "Correlation with PC1 and PC2",bg="tomato",lcolor="tomato",cex=1.1, xlim=range(-1,1))
  abline(v = 0, lty = 3)
  
  
  # PC 3 and PC4
  dotchart(t(alignPCAcorrelations[3:4,1:ncol(alignPCAcorrelations)]), xlab = "correlation estimate", main = "Correlation with PC3 and PC4",bg="tomato",lcolor="tomato",cex=1.1, xlim=range(-1,1))
  abline(v = 0, lty = 3)
}
```


Results from multiple sailfish outputs
```{r}
align$sampNames = lookupRev$UDF.Investigator.Sample.Name[match(as.character(align$SAMPLE), lookupRev$Sample.Name)]
head(align)
metricsToCheck = c('MEDIAN_3PRIME_BIAS', 'MEDIAN_5PRIME_BIAS', 'MEDIAN_5PRIME_TO_3PRIME_BIAS', 'MEDIAN_CV_COVERAGE', 'PCT_CODING_BASES', 'PCT_CORRECT_STRAND_READS', 'PCT_INTEGENIC_BASES', 'PCT_INTRONIC_BASES', 'PCT_MRNA_BASES', 'PCT_USABLE_BASES', 'PCT_UTR_BASES')
alignToCheck = align[,which(colnames(align) %in% metricsToCheck)]

op = par(mfrow = c(3,3))
for (i in 1:ncol(alignToCheck)){ hist(alignToCheck[,i], main = colnames(alignToCheck)[i])}
par(op)
library(corrgram)
corrgram(alignToCheck,lower.panel = panel.pts,upper.panel = panel.pie)

plotcols = c("orangered", "royalblue")

pcaPlots=function(in.dge,pcNames){
  par(op)
  pc_palo.dge = generate_PCA_datasets(in.dge = in.dge, pcNames=pcNames)
  pc_palx20.pca = prcomp_PCA(filterByFractionPresent(inDGE = pc_palo.dge,fraction = 0.2,minCount = 1))
  #biplot(pc_palx20.pca,var.axes = FALSE)
  # Factomine plot of samples
  caseStatus = as.character(pc_palo.dge$samples$group)
  caseStatus[which(clinicalDataR$"Sample Type" == "Control")] = "control"
  fm20.pca = FM_PCA(filterByFractionPresent(inDGE = pc_palo.dge,fraction = 0.2,minCount = 1))
  plot.PCA(fm20.pca, choix = "ind", new.plot = TRUE,col.ind = plotcols[as.numeric(as.factor(caseStatus))],label = "none")
  legend("topright",legend = c("case", "control"),col = plotcols,pch = 16)
  calc_plot_PC_corr(in.pca = pc_palx20.pca$x,inClinical = clinicalDataR)
  calc_plot_alignPC_corr(in.prc = pc_palx20.pca$x,inAlign = alignToCheck)
  par(op)
}

##################
# Estimated reads
tmp = generateDataAndClinicalObj('syn3064890')
data.dge = tmp$DGE
clinicalDataR = tmp$VARS
head(clinicalDataR)
clinicalDataR = clinicalDataR[match(colnames(data.dge),clinicalDataR$"Px Code"),]
temp = clinicalDataR
toConvert = c(8,9,12,13,14)
for (i in toConvert){
  temp[,i] = as.numeric(temp[,i])
}
corrgram(temp,lower.panel = panel.pts,upper.panel = panel.pie)
rm(temp)
pcaPlots(in.dge=data.dge,pcNames=proteinCoding_transcripts)


##################
# Estimate Kmers
plot(1, main = "estimated Kmers")
tmp = generateDataAndClinicalObj('syn3098045')
data.dge = tmp$DGE
clinicalDataR = tmp$VARS
head(clinicalDataR)
clinicalDataR = clinicalDataR[match(colnames(data.dge),clinicalDataR$"Px Code"),]
pcaPlots(in.dge=data.dge,pcNames=proteinCoding_transcripts)

#########
## TPM ##
# Using slighly different code, since this data is already TPM and doesn't need cpm transformation
plot(1, main = "TPM")
dataEntity = synGet('syn3064954') 
library('R.utils')
gunzip(getFileLocation(dataEntity),overwrite=TRUE)
x = unlist(strsplit(getFileLocation(dataEntity), split = '.gz'))
detach("package:R.utils", unload=TRUE)
estReads = read.csv(x, row.names = 1)
colnames(estReads) = lookupRev$UDF.Investigator.Sample.Name[match(colnames(estReads), paste("X", lookupRev$Sample.Name,sep = ""))]


data.dge = DGEList(counts=estReads,group=factor(casesStatus),remove.zeros=TRUE)
b = as.list(rownames(data.dge))
strippedNames = sapply(b,function(x){unlist(strsplit(as.character(x), split = "[.]"))[1]})
tmp.dge = data.dge[which(strippedNames %in% proteinCoding_transcripts[,1]),]
pc.dge = DGEList(getCounts(tmp.dge),group = tmp.dge$samples$group)
caseStatus = as.character(pc.dge$samples$group)
caseStatus[which(clinicalDataR$"Sample Type" == "Control")] = "control"

head(clinicalDataR)
clinicalDataR = clinicalDataR[match(colnames(pc.dge),clinicalDataR$"Px Code"),]
tpm_pc.prc = prcomp(t(getCounts(pc.dge)),center=TRUE,scale=TRUE,retx = TRUE)
barplot((tpm_pc.prc$sdev^2/sum(tpm_pc.prc$sdev^2))[1:20], ylab = "fraction variance") 
# Factomine plot of samples
tpm_fm20.pca = PCA(X = t(getCounts(pc.dge)),scale.unit = TRUE,ncp = 5,graph = FALSE)
barplot(tpm_fm20.pca$eig$"percentage of variance"[1:20], ylab = "% variance") 
plot.PCA(tpm_fm20.pca, choix = "ind", new.plot = TRUE,col.ind = plotcols[as.numeric(as.factor(caseStatus))],label = "none")
legend("topright",legend = c("case", "control"),col = plotcols,pch = 16)
par(op)
calc_plot_PC_corr(in.pca = tpm_pc.prc$x,inClinical = clinicalDataR)
calc_plot_alignPC_corr(in.prc = tpm_pc.prc$x,inAlign = alignToCheck)
par(op)

#################
## Gene counts ##
plot(1, main = "gene counts")
par(mai = c(1.991176, 1.600749, 1.600749, 0.819896))
tmp = generateDataAndClinicalObj('syn2820309')
data.dge = tmp$DGE
clinicalDataR = tmp$VARS
head(clinicalDataR)
clinicalDataR = clinicalDataR[match(colnames(data.dge),clinicalDataR$"Px Code"),]
pcaPlots(in.dge=data.dge,pcNames=proteinCoding_genes)
```
