---
title: "swap_mapping_metrics.Rmd"
author: "Kristen Dang"
date: "October 6, 2015"
output: 
  html_document: 
    fig_height: 7
---

```{r}
library('githubr')
source('/Users/kkdang/Computing/rgithubclient_authenticate.R')
library(edgeR)
library(gplots)
library(RColorBrewer)
library(beeswarm)
sourceRepoFile(sageCode, "rnaseq_analysis_functions.R")
sourceRepoFile(sageCode, "NIA-AMP-AD/swap_analysis/mayo_swap_preliminary.R")
```

# Alignment metrics
```{r}
starAlignMetrics = synTableQuery("SELECT * FROM syn4584502")
colnames(starAlignMetrics@values)

starAlignMetrics@values$Sample
adjSampleNames = sapply(strsplit(as.character(starAlignMetrics@values$Sample), "[.]"), function(x){ return(x[1]) } )
toKeep = which(paste("X",adjSampleNames,sep="") %in% c(colnames(mssmData), colnames(mayoData)))
plotdata = starAlignMetrics@values[toKeep,]
plotdata$site = c(rep("mssm",9),rep("mayo",9))


beeswarm(`Uniq-reads-PCT`~site,data=plotdata, ylab = "Uniquely mapped %", col = c("goldenrod","red"),pch = 16)
beeswarm(`Number-reads`~site,data=plotdata, ylab = "Number of reads", col = c("goldenrod","red"),pch = 16)
beeswarm(`Ave-read-length`~site,data=plotdata, ylab = "read length", col = c("goldenrod","red"),pch = 16)
beeswarm(`Reads-mapped-to-multiple-loci-PCT`~site,data=plotdata, ylab = "Multiple loci %", col = c("goldenrod","red"),pch = 16)
beeswarm(`Reads-unmapped-other-PCT`~site,data=plotdata, ylab = "Unmapped, not mismatch %", col = c("goldenrod","red"),pch = 16)
beeswarm(`Reads-unmapped-too-short-PCT`~site,data=plotdata, ylab = "Unmapped, too short %", col = c("goldenrod","red"),pch = 16)
beeswarm(`Total-splices`~site,data=plotdata, ylab = "Total splices", col = c("goldenrod","red"),pch = 16)

```



# Picard gene model metrics
```{r}
picardGeneMetrics = synTableQuery("SELECT * FROM syn4597336")
colnames(picardGeneMetrics@values)

picardGeneMetrics@values$Sample
adjSampleNames = sapply(strsplit(as.character(picardGeneMetrics@values$Sample), "[.]"), function(x){ return(x[1]) } )
toKeep = which(paste("X",adjSampleNames,sep = "") %in% c(colnames(mssmData), colnames(mayoData)))
plotdata = picardGeneMetrics@values[toKeep,]
plotdata$site = c(rep("mssm",9),rep("mayo",9))


plot(plotdata$PCT_INTERGENIC_BASES,col = "plum4",pch = 16, xlab = "samples",type = "o",ylab = "fraction reads", ylim = range(0,1))
lines(plotdata$PCT_INTRONIC_BASES,col = "slateblue",pch = 16, xlab = "samples",type = "o")
lines(plotdata$PCT_UTR_BASES,col = "seagreen3",pch = 16, xlab = "samples",type = "o")
lines(plotdata$PCT_CODING_BASES,col = "darkgoldenrod2",pch = 16, xlab = "samples",type = "o")
lines(plotdata$PCT_RIBOSOMAL_BASES,col = "orangered",pch = 16, xlab = "samples",type = "o")
legend("topright",legend = c("intergenic","intronic","UTR","coding","ribosomal"),col = c("plum4","slateblue","seagreen3","darkgoldenrod2","orangered"),pch=16)

```


# featurecounts assignment metrics
```{r}
fcAssignMetrics = synTableQuery("SELECT * FROM syn4601557")
colnames(fcAssignMetrics@values)

adjSampleNames = sapply(strsplit(as.character(fcAssignMetrics@values$Sample), "[.]"), function(x){ return(x[1]) } )
toKeep = which(paste("X", adjSampleNames,sep = "") %in% c(colnames(mssmData),"766_TCX"))
plotdata = fcAssignMetrics@values[toKeep,]


z = substr(plotdata$Sample,start = 1,stop = 8)
tmp = t(data.matrix(plotdata[,3:7]))
colnames(tmp) = z
tmp = tmp[,order(colnames(tmp))]

barplot(tmp,beside = FALSE,las = 2,legend.text = colnames(plotdata[,3:7]),col = brewer.pal(n=5,name="Accent"),main = "read counting results")

```
