---
title: "swap_mapping_metrics.Rmd"
author: "Kristen Dang"
date: "October 6, 2015"
output: 
  html_document: 
    fig_height: 7
---

```{r}
library('rGithubClient')
source('/Users/kristen/Computing/external_software/rgithubclient_authenticate.R')
library(edgeR)
library(gplots)
library(RColorBrewer)
sourceRepoFile(sageCode, "rnaseq_analysis_functions.R")
sourceRepoFile(sageCode, "NIA-AMP-AD/swap_analysis/sample_swap_preliminary.R")
```


# Trim metrics
```{r}
fastqTrimMetrics = synTableQuery("SELECT * FROM syn4616136")
colnames(fastqTrimMetrics@values)
plot(fastqTrimMetrics@values$R1_num_empty_reads/fastqTrimMetrics@values$R1_num_reads, ylab = "R1 fraction to short after trim", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(fastqTrimMetrics@values$R1_num_trimmed_bases, ylab = "R1 fraction trimmed bases", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(fastqTrimMetrics@values$R2_num_empty_reads/fastqTrimMetrics@values$R2_num_reads, ylab = "R2 fraction to short after trim", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(fastqTrimMetrics@values$R2_num_trimmed_bases, ylab = "R2 fraction trimmed bases", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
```

# Alignment metrics
```{r}
starAlignMetrics = synTableQuery("SELECT * FROM syn4584502")
colnames(starAlignMetrics@values)
plot(starAlignMetrics@values$`Uniq-reads-PCT`, ylab = "Uniquely mapped", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topleft",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(starAlignMetrics@values$`Number-reads`, ylab = "Number of reads", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(starAlignMetrics@values$`Ave-read-length`, ylab = "read length", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(starAlignMetrics@values$`Reads-mapped-to-multiple-loci-PCT`, ylab = "Multiple loci", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(starAlignMetrics@values$`Reads-unmapped-other-PCT`, ylab = "Unmapped, not mismatch", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topleft",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(starAlignMetrics@values$`Reads-unmapped-too-short-PCT`, ylab = "Unmapped, too short", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
plot(starAlignMetrics@values$`Total-splices`, ylab = "Total splices", col = c(rep("goldenrod",10),rep("red",10),rep("blue",10)),pch = 16)
legend("topright",legend = c("Mayo","Broad","MSSM"),col = c("goldenrod","red","blue"),pch=16)
```

# Picard gene model metrics
```{r}
picardGeneMetrics = synTableQuery("SELECT * FROM syn4597336")
colnames(picardGeneMetrics@values)
picardGeneMetrics@values$Sample
plotdata = picardGeneMetrics@values[1:30,]


mayoAveResults = colMeans(plotdata[1:10,3:24])
mssmAveResults = colMeans(plotdata[21:30,3:24])
broadAveResults = colMeans(plotdata[11:20,3:24])

plot(plotdata$PCT_INTERGENIC_BASES,col = "plum4",pch = 16, xlab = "samples",type = "o",ylab = "fraction reads", ylim = range(0,1))
lines(plotdata$PCT_INTRONIC_BASES,col = "slateblue",pch = 16, xlab = "samples",type = "o")
lines(plotdata$PCT_UTR_BASES,col = "seagreen3",pch = 16, xlab = "samples",type = "o")
lines(plotdata$PCT_CODING_BASES,col = "darkgoldenrod2",pch = 16, xlab = "samples",type = "o")
lines(plotdata$PCT_RIBOSOMAL_BASES,col = "orangered",pch = 16, xlab = "samples",type = "o")
legend("topright",legend = c("intergenic","intronic","UTR","coding","ribosomal"),col = c("plum4","slateblue","seagreen3","darkgoldenrod2","orangered"),pch=16)

barplot(t(rbind(mayoAveResults[3:7]/mayoAveResults[2], broadAveResults[3:7]/broadAveResults[2], mssmAveResults[3:7]/mssmAveResults[2])),beside = FALSE,names.arg = c("siteA", "siteB", "siteC"),legend.text = c("ribosomal", "coding", "UTR", "intronic", "intergenic"), col = rainbow(5), ylab = "fraction of aligned reads")
```

# featurecounts assignment metrics
```{r}
fcAssignMetrics = synTableQuery("SELECT * FROM syn4601557")
colnames(fcAssignMetrics@values)
barplot(t(data.matrix(fcAssignMetrics@values[,3:7])),beside = FALSE,las = 2,names.arg = fcAssignMetrics@values$Sample,legend.text = colnames(fcAssignMetrics@values)[3:7],col = brewer.pal(n=5,name="Accent"),main = "read counting results")
```
