---
title: "swap_summary.Rmd"
author: "Kristen Dang"
date: "May 27, 2016"
output: html_document
---

```{r}
library(synapseClient)
synapseLogin()
library('githubr')
source('/Users/kkdang/Computing/rgithubclient_authenticate.R')
#sourceRepoFile(sageCode, "NIA-AMP-AD/sample_swap_preliminary.R")
```


# Alignment metrics
```{r}
starAlignMetrics = synTableQuery("SELECT * FROM syn4584502")
colnames(starAlignMetrics@values)

metricsNames = sapply(as.list(starAlignMetrics@values$Sample), function(x) {unlist(strsplit(x, split = "[.]"))[1]})

TCX = metricsNames[grep(metricsNames,pattern = "TCX")]
notSwap = setdiff(paste("X",TCX, sep = ""), c(Bsamples, Ysamples, Msamples))

toRemove = c(which(paste("X", starAlignMetrics@values$Sample, sep = "") %in% paste(notSwap, ".snap", sep = "")), which( starAlignMetrics@values$Sample %in% "05689621.snap"))

swapData = starAlignMetrics@values[-toRemove,]


#beeswarm(`Uniq-reads-PCT`~SeqSite+StudyOrigin, data=swapData, ylab = "Uniquely mapped", col = c("blue", "goldenrod","red"),pch = 16, las = 2, xlab = "",labels = c("", rep("Mayo", 2), "", rep("MSSM", 2), rep("ROSMAP",3)))
boxplot(`Uniq-reads-PCT`~SeqSite+StudyOrigin, data=swapData, main = "Uniquely mapped reads", ylab = "percent", col = c("blue", "goldenrod","red"),pch = 16, las = 2, xlab = "",names = c("", rep("Mayo", 2), "", rep("MSSM", 2), rep("ROSMAP",3)), varwidth = TRUE)
legend("topleft",legend = c("Mayo   ","MSSM   ", "Broad   "),col = c("goldenrod","red","blue"),pch=16)


#beeswarm(`Number-reads`~SeqSite+StudyOrigin, data=swapData, ylab = "", main = "Number of reads", col = c("blue", "goldenrod","red"),pch = 16, las = 2, xlab = "",labels = c("", rep("Mayo", 2), "", rep("MSSM", 2), rep("ROSMAP",3)))
boxplot(`Number-reads`~SeqSite+StudyOrigin, data=swapData, main = "Number of reads", col = c("blue", "goldenrod","red"),pch = 16, las = 2, xlab = "",names = c("", rep("Mayo", 2), "", rep("MSSM", 2), rep("ROSMAP",3)), varwidth = TRUE)
legend("topleft",legend = c("Mayo   ","MSSM   ", "Broad   "),col = c("goldenrod","red","blue"),pch=16)

```


# Picard gene model metrics
```{r}
# picardGeneMetrics = synTableQuery("SELECT * FROM syn4597336")
# colnames(picardGeneMetrics@values)
# picardGeneMetrics@values$Sample
# plotdata = picardGeneMetrics@values[1:30,]

op = par(mfrow = c(1,3))
picardGeneMetricsFile = synTableQuery("SELECT * FROM syn4597336",loadResult = FALSE)
picardGeneMetrics = read.csv(getFileLocation(picardGeneMetricsFile))
colnames(picardGeneMetrics)
picardGeneMetrics$Sample

# From Mayo
picardNames = sapply(as.list(as.character(picardGeneMetrics$Sample)), function(x) {unlist(strsplit(x, split = "[.]"))[1]})
temp = picardGeneMetrics[grep(pattern = "TCX", x = picardNames),3:ncol(picardGeneMetrics)]
plotdata = temp[-which(paste("X", temp$Sample, sep = "") %in% paste(notSwap, ".snap", sep = "")),]

YmssmAveResults = colMeans(plotdata[1:9,3:24])
YmayoAveResults = colMeans(plotdata[10:18,3:24])

barplot(t(rbind(YmayoAveResults[3:7]/YmayoAveResults[2], YmssmAveResults[3:7]/YmssmAveResults[2])),beside = FALSE,names.arg = c("Mayo", "MSSM"),legend.text = c("ribosomal", "coding", "UTR", "intronic", "intergenic"), col = rainbow(5), ylab = "fraction of aligned reads", main = "Mayo")

# From MSSM
plotdata = picardGeneMetrics[(nrow(picardGeneMetrics)-18):nrow(picardGeneMetrics),3:ncol(picardGeneMetrics)]

MmayoAveResults = colMeans(plotdata[1:9,3:24])
MmssmAveResults = colMeans(plotdata[10:19,3:24])

barplot(t(rbind(MmayoAveResults[3:7]/MmayoAveResults[2], MmssmAveResults[3:7]/MmssmAveResults[2])),beside = FALSE,names.arg = c("Mayo", "MSSM"),legend.text = c("ribosomal", "coding", "UTR", "intronic", "intergenic"), col = rainbow(5), ylab = "fraction of aligned reads", main = "MSSM")


# From ROSMAP
plotdata = picardGeneMetrics[1:30,3:ncol(picardGeneMetrics)]

RmayoAveResults = colMeans(plotdata[1:10,3:24])
RmssmAveResults = colMeans(plotdata[21:30,3:24])
RbroadAveResults = colMeans(plotdata[11:20,3:24])

barplot(t(rbind(RmayoAveResults[3:7]/RmayoAveResults[2], RbroadAveResults[3:7]/RbroadAveResults[2], RmssmAveResults[3:7]/RmssmAveResults[2])),beside = FALSE,names.arg = c("Mayo", "Broad", "MSSM"),legend.text = c("ribosomal", "coding", "UTR", "intronic", "intergenic"), col = rainbow(5), ylab = "fraction of aligned reads", main = "ROSMAP")
par(op)
```


# Effective depth (reads covering coding regions)
```{r}
op = par(mfrow = c(1,3))

# From Mayo
picardNames = sapply(as.list(as.character(picardGeneMetrics$Sample)), function(x) {unlist(strsplit(x, split = "[.]"))[1]})
temp = picardGeneMetrics[grep(pattern = "TCX", x = picardNames),3:ncol(picardGeneMetrics)]
plotdata = temp[-which(paste("X", temp$Sample, sep = "") %in% paste(notSwap, ".snap", sep = "")),]
boxplot(plotdata$CODING_BASES[10:18]/200, plotdata$CODING_BASES[1:9]/100, main = "Mayo", col = c("goldenrod","red", "blue"),pch = 16, las = 2, xlab = "",names = c("Mayo", "MSSM"), varwidth = TRUE, ylim = range(5e6, 3.5e7))


# From MSSM
plotdata = picardGeneMetrics[(nrow(picardGeneMetrics)-18):nrow(picardGeneMetrics),3:ncol(picardGeneMetrics)]
boxplot(plotdata$CODING_BASES[1:9]/200, plotdata$CODING_BASES[10:19]/100,  main = "MSSM", col = c("goldenrod","red", "blue"),pch = 16, las = 2, xlab = "",names = c("Mayo", "MSSM"), varwidth = TRUE,  ylim = range(5e6, 3.5e7))


# From ROSMAP
plotdata = picardGeneMetrics[1:30,3:ncol(picardGeneMetrics)]
boxplot(plotdata$CODING_BASES[1:10]/200, plotdata$CODING_BASES[21:30]/100, plotdata$CODING_BASES[11:20]/200,  main = "ROSMAP", col = c("goldenrod","red", "blue"),pch = 16, las = 2, xlab = "",names = c("Mayo", "MSSM","Broad"), varwidth = TRUE,  ylim = range(5e6, 3.5e7))

par(op)
```


Stability of quant across depths
```{r}

downSampleCounts=function(countVec, frac = 0.5){
  zeroPos = which(countVec == 0)
  countVecNoZero = countVec[-zeroPos]
  countVecOut = countVecNoZero
  runningSum = cumsum(countVecNoZero)
  
  probToRemove = 1-frac
  numberToRemove = probToRemove * sum(countVec)
  countsToRemove = sample.int(n = sum(countVec), size = numberToRemove,replace = FALSE)
  
  posToMutate = sapply(as.list(countsToRemove), function(x){which((x > runningSum) == FALSE)[1]})
  
  x = as.data.frame(table(posToMutate))
  xPos = as.numeric(as.character(x$posToMutate))
  countVecOut[xPos] = countVecNoZero[xPos] - x$Freq

  countVecNoZero.cpm = cpm(countVecNoZero,log = FALSE)
  countVecOut.cpm = cpm(countVecOut,log = FALSE)
  
  return((countVecNoZero.cpm-countVecOut.cpm)/countVecNoZero.cpm)
#  return(countVecOut.cpm)
#  return(countVecOut)
}
# test
countVec = c(0,34,12,0,1, 102, 4, 9, 14)
test1 = downSampleCounts(countVec = countVec)
boxplot(test1)

head(YmssmPalx.dge$counts)
colSums(YmssmPalx.dge$counts)
hist(colSums(YmssmPalx.dge$counts))

# downsample counts from 6929_TCX
which(colnames(YmssmPalx.dge$counts) == "X6929_TCX")
YM_6929 = YmssmPalx.dge$counts[,3]
YM_6929_9 = downSampleCounts(countVec = YM_6929,frac = 0.9)
YM_6929_7 = downSampleCounts(countVec = YM_6929,frac = 0.7)
YM_6929_5 = downSampleCounts(countVec = YM_6929,frac = 0.5)
YM_6929_3 = downSampleCounts(countVec = YM_6929,frac = 0.3)


YM_6929_noZero = YM_6929[-which(YM_6929 == 0)]
YM_6929.dge = DGEList(counts = YM_6929_noZero)
YM_6929.cpm = cpm(YM_6929.dge,log = FALSE)


boxplot(YM_6929_3)
dsCounts = YM_6929_noZero - (YM_6929_3 * YM_6929_noZero)
YM_6929_3.dge = DGEList(counts = dsCounts,remove.zeros = FALSE)
YM_6929_3.cpm = cpm(YM_6929_3.dge,log = FALSE)
boxplot((YM_6929.cpm - YM_6929_3.cpm)/YM_6929.cpm)




# downsample 10% of sample 6929_TCX
toKeep = sample.int(n = length(YM_6929), size = ceiling(0.1*length(YM_6929)),replace = FALSE)
YM_6929_10 = YM_6929[toKeep]
YM_6929_9_10 = downSampleCounts(countVec = YM_6929_10,frac = 0.9)
YM_6929_7_10 = downSampleCounts(countVec = YM_6929_10,frac = 0.7)
YM_6929_5_10 = downSampleCounts(countVec = YM_6929_10,frac = 0.5)
YM_6929_3_10 = downSampleCounts(countVec = YM_6929_10,frac = 0.3)
boxplot(YM_6929_9_10[,1], YM_6929_7_10[,1], YM_6929_5_10[,1], YM_6929_3_10[,1],names = c("90%", "70%", "50%", "30%"),outline = FALSE)
boxplot(YM_6929_9_10[,1], YM_6929_7_10[,1], YM_6929_5_10[,1], YM_6929_3_10[,1],names = c("90%", "70%", "50%", "30%"),outline = TRUE)

# from 30% of genes
toKeep = sample.int(n = length(YM_6929), size = ceiling(0.3*length(YM_6929)),replace = FALSE)
YM_6929_30 = YM_6929[toKeep]
YM_6929_9_30 = downSampleCounts(countVec = YM_6929_30,frac = 0.9)
YM_6929_7_30 = downSampleCounts(countVec = YM_6929_30,frac = 0.7)
YM_6929_5_30 = downSampleCounts(countVec = YM_6929_30,frac = 0.5)
YM_6929_3_30 = downSampleCounts(countVec = YM_6929_30,frac = 0.3)
boxplot(YM_6929_9_30[,1], YM_6929_7_30[,1], YM_6929_5_30[,1], YM_6929_3_30[,1],names = c("90%", "70%", "50%", "30%"),outline = FALSE, main = "fractional difference in normalized count", ylab = "diff/exp", xlab = "downsampled fraction")
boxplot(YM_6929_9_30[,1], YM_6929_7_30[,1], YM_6929_5_30[,1], YM_6929_3_30[,1],names = c("90%", "70%", "50%", "30%"),outline = TRUE, main = "fractional difference in normalized count", ylab = "diff/exp", xlab = "downsampled fraction")

```