---
title: "swap_summary.Rmd"
author: "Kristen Dang"
date: "May 27, 2016"
output: html_document
---

```{r}
library(synapseClient)
synapseLogin()
library('githubr')
source('/Users/kkdang/Computing/rgithubclient_authenticate.R')
#sourceRepoFile(sageCode, "NIA-AMP-AD/sample_swap_preliminary.R")
```


# Alignment metrics
```{r}
starAlignMetrics = synTableQuery("SELECT * FROM syn4584502")
colnames(starAlignMetrics@values)

metricsNames = sapply(as.list(starAlignMetrics@values$Sample), function(x) {unlist(strsplit(x, split = "[.]"))[1]})

TCX = metricsNames[grep(metricsNames,pattern = "TCX")]
notSwap = setdiff(paste("X",TCX, sep = ""), c(Bsamples, Ysamples, Msamples))

toRemove = c(which(paste("X", starAlignMetrics@values$Sample, sep = "") %in% paste(notSwap, ".snap", sep = "")), which( starAlignMetrics@values$Sample %in% "05689621.snap"))

swapData = starAlignMetrics@values[-toRemove,]


#beeswarm(`Uniq-reads-PCT`~SeqSite+StudyOrigin, data=swapData, ylab = "Uniquely mapped", col = c("blue", "goldenrod","red"),pch = 16, las = 2, xlab = "",labels = c("", rep("Mayo", 2), "", rep("MSSM", 2), rep("ROSMAP",3)))
boxplot(`Uniq-reads-PCT`~SeqSite+StudyOrigin, data=swapData, ylab = "Uniquely mapped fraction", col = c("blue", "goldenrod","red"),pch = 16, las = 2, xlab = "",names = c("", rep("Mayo", 2), "", rep("MSSM", 2), rep("ROSMAP",3)), varwidth = TRUE)
legend("topleft",legend = c("Mayo   ","MSSM   ", "Broad   "),col = c("goldenrod","red","blue"),pch=16)


#beeswarm(`Number-reads`~SeqSite+StudyOrigin, data=swapData, ylab = "", main = "Number of reads", col = c("blue", "goldenrod","red"),pch = 16, las = 2, xlab = "",labels = c("", rep("Mayo", 2), "", rep("MSSM", 2), rep("ROSMAP",3)))
boxplot(`Number-reads`~SeqSite+StudyOrigin, data=swapData, main = "Number of reads", col = c("blue", "goldenrod","red"),pch = 16, las = 2, xlab = "",names = c("", rep("Mayo", 2), "", rep("MSSM", 2), rep("ROSMAP",3)), varwidth = TRUE)
legend("topleft",legend = c("Mayo   ","MSSM   ", "Broad   "),col = c("goldenrod","red","blue"),pch=16)

```


# Picard gene model metrics
```{r}
# picardGeneMetrics = synTableQuery("SELECT * FROM syn4597336")
# colnames(picardGeneMetrics@values)
# picardGeneMetrics@values$Sample
# plotdata = picardGeneMetrics@values[1:30,]

op = par(mfrow = c(1,3))
picardGeneMetricsFile = synTableQuery("SELECT * FROM syn4597336",loadResult = FALSE)
picardGeneMetrics = read.csv(getFileLocation(picardGeneMetricsFile))
colnames(picardGeneMetrics)
picardGeneMetrics$Sample
plotdata = picardGeneMetrics[1:30,3:ncol(picardGeneMetrics)]

RmayoAveResults = colMeans(plotdata[1:10,3:24])
RmssmAveResults = colMeans(plotdata[21:30,3:24])
RbroadAveResults = colMeans(plotdata[11:20,3:24])

barplot(t(rbind(RmayoAveResults[3:7]/RmayoAveResults[2], RbroadAveResults[3:7]/RbroadAveResults[2], RmssmAveResults[3:7]/RmssmAveResults[2])),beside = FALSE,names.arg = c("Mayo", "Broad", "MSSM"),legend.text = c("ribosomal", "coding", "UTR", "intronic", "intergenic"), col = rainbow(5), ylab = "fraction of aligned reads", main = "ROSMAP")

# From Mayo
picardNames = sapply(as.list(as.character(picardGeneMetrics$Sample)), function(x) {unlist(strsplit(x, split = "[.]"))[1]})
temp = picardGeneMetrics[grep(pattern = "TCX", x = picardNames),3:ncol(picardGeneMetrics)]
plotdata = temp[-which(paste("X", temp$Sample, sep = "") %in% paste(notSwap, ".snap", sep = "")),]

YmssmAveResults = colMeans(plotdata[1:9,3:24])
YmayoAveResults = colMeans(plotdata[10:18,3:24])

barplot(t(rbind(YmayoAveResults[3:7]/YmayoAveResults[2], YmssmAveResults[3:7]/YmssmAveResults[2])),beside = FALSE,names.arg = c("Mayo", "MSSM"),legend.text = c("ribosomal", "coding", "UTR", "intronic", "intergenic"), col = rainbow(5), ylab = "fraction of aligned reads", main = "Mayo")

# From MSSM
plotdata = picardGeneMetrics[(nrow(picardGeneMetrics)-18):nrow(picardGeneMetrics),3:ncol(picardGeneMetrics)]

MmayoAveResults = colMeans(plotdata[1:9,3:24])
MmssmAveResults = colMeans(plotdata[10:19,3:24])

barplot(t(rbind(MmayoAveResults[3:7]/MmayoAveResults[2], MmssmAveResults[3:7]/MmssmAveResults[2])),beside = FALSE,names.arg = c("Mayo", "MSSM"),legend.text = c("ribosomal", "coding", "UTR", "intronic", "intergenic"), col = rainbow(5), ylab = "fraction of aligned reads", main = "MSSM")

```


# Effective depth
```{r}


```